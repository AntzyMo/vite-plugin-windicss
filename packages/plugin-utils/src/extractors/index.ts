export * from './default'
export * from './helper'
export * from './pug'
export * from './svelte'
