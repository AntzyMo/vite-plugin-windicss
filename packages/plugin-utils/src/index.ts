export * from './constants'
export * from './defineConfig'
export * from './createUtils'
export * from './options'
export * from './transforms'
export * from './extractors'
export * from './utils'
export * from './resolveOptions'
export * from './merge'
