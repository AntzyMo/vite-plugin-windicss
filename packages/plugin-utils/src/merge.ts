import type { FullConfig } from 'windicss/types/interfaces'

const isObject = (val: any) => toString.call(val) === '[object Object]'

function deepMerge(a: any, b: any, rootPath: string) {
  a = { ...a }
  Object.keys(b).forEach((key) => {
    if (isObject(a[key]))
      a[key] = deepMerge(a[key], b[key], rootPath ? `${rootPath}.${key}` : key)
    else if (Array.isArray(a[key]))
      a[key] = [...a[key], ...b[key]]
    else
      a[key] = b[key]
  })
  return a
}

export function mergeWindicssConfig(a: FullConfig, b: FullConfig) {
  // TODO: handle more special props
  return deepMerge(a, b, '')
}
