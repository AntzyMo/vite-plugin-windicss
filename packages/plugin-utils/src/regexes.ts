export const regexQuotedString = /(["'`])((?:\\\1|(?:(?!\1)|\n|\r).)*?)\1/mg
export const regexClassCheck = /^[!a-z\d.+-]+[\w:/\\.$()-]*[\w)]$/
export const regexHtmlTag = /<(\w[\w-]*)[\S\s]*?\/?>/mg
export const regexClassSplitter = /[\s'"`{}]/g
export const regexClassGroup = /([!\w+-][\w+:_/-]*?\w):\(([\w\s/-]*?)\)/gm
