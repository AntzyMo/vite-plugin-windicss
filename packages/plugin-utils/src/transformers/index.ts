export * from './pug'
export * from './types'
