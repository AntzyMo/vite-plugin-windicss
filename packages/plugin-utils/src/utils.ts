import { toArray } from '@antfu/utils'

export type { Arrayable } from '@antfu/utils'
export { toArray, partition, slash } from '@antfu/utils'

export type NestedArrayable<T> = T | (T | T[])[]

export function flattenArray<T>(v: NestedArrayable<T>): T[] {
  return toArray(v).flat() as T[]
}

export function mergeArrays<T>(...args: (NestedArrayable<T> | undefined)[]): T[] {
  return args.flatMap(i => flattenArray(i || []))
}

export function kebabCase(str: string) {
  return str.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase()
}

export function include<T>(set: Set<T>, v: T[] | Set<T>) {
  for (const i of v)
    set.add(i)
}

export function exclude<T>(set: Set<T>, v: T[] | Set<T>) {
  for (const i of v)
    set.delete(i)
}

export function escapeRegExp(str: string) {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&') // $& means the whole matched string
}
