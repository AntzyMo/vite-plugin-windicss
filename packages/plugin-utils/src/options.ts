/* eslint-disable no-use-before-define */

import type { Config as WindiCssOptions } from 'windicss/types/interfaces'
import { defaultAlias, TagNames } from './constants'
import { kebabCase, toArray } from './utils'

export { WindiCssOptions }

export interface WindiPluginUtilsOptions {
  /**
   * Options for windicss/tailwindcss.
   * Also accepts string as config file path.
   *
   * @default 'tailwind.config.js'
   */
  config?: WindiCssOptions | string

  /**
   * Safe class names to be always included.
   */
  safelist?: string | string[]

  /**
   * Enabled windicss preflight (a.k.a TailwindCSS style reset)
   *
   * @default true
   */
  preflight?: boolean | {
    /**
     * Enable all the preflight regardless the template
     */
    enableAll?: boolean

    /**
     * Safelist to always included
     */
    safelist?: string | string[]

    /**
      * Alias for resolving preflight
      */
    alias?: Record<string, TagNames>

    /**
     * @default true
     */
    includeBase?: boolean

    /**
     * @default true
     */
    includeGlobal?: boolean

    /**
     * @default true
     */
    includePlugin?: boolean
  }

  /**
   * Scan the files and extract the usage
   *
   * @default true
   */
  scan?: boolean | {
    /**
     * Auto scan on startup
     *
     * @default true
     */
    runOnStartup?: boolean

    /**
     * Directories to search for classnames
     *
     * @default 'src'
     */
    dirs?: string[]

    /**
     * File extension to search for classnames
     *
     * @default 'html', 'vue'
     */
    fileExtensions?: string[]

    /**
     * Exclude globs
     *
     * @default []
     */
    exclude?: string[]

    /**
     * Include globs
     *
     * @default []
     */
    include?: string[]
  }

  /**
   * Transform CSS for `@apply` directive
   *
   * @default true
   */
  transformCSS?: boolean

  /**
   * Transform groups like `hover:(bg-gray-100 font-medium)`
   *
   * @default true
   */
  transformGroups?: boolean

  /**
   * Sort the genrate utilities
   *
   * @default true
   */
  sortUtilities?: boolean

  /**
   * Name for debug
   *
   * @default 'windi-plugin-utils'
   * @internal
   */
  _pluginName?: string

  /**
   * CWD
   *
   * @default process.cwd
   * @internal
   */
  _projectRoot?: string
}

export type UserOptions = Omit<WindiPluginUtilsOptions, '_pluginName' | '_projectRoot'>

export function resolveOptions(options: WindiPluginUtilsOptions) {
  const {
    config = 'tailwind.config.js',
    scan = true,
    preflight = true,
    transformCSS = true,
    sortUtilities = true,
    _projectRoot = process.cwd(),
    _pluginName = 'windi-plugin-utils',
  } = options

  const preflightOptions = Object.assign(
    {
      includeBase: true,
      includeGlobal: true,
      includePlugin: true,
      enableAll: false,
      alias: {},
    },
    typeof preflight === 'boolean' ? {} : preflight,
  )

  const scanOptions = Object.assign(
    {
      fileExtensions: ['html', 'vue', 'md', 'pug', 'jsx', 'tsx', 'svelte'],
      dirs: ['src'],
      exclude: [] as string[],
      include: [] as string[],
      runOnStartup: true,
    },
    typeof scan === 'boolean' ? {} : scan,
  )

  const safelist = new Set(toArray(options.safelist || []).flatMap(i => i.split(' ')))

  preflightOptions.alias = Object.fromEntries(
    Object.entries({
      ...defaultAlias,
      ...preflightOptions.alias,
    }).filter(([k, v]) => [kebabCase(k), v]),
  )

  return {
    ...options,
    _projectRoot,
    _pluginName,
    config,
    scan: Boolean(scan),
    scanOptions,
    preflight: Boolean(preflight),
    preflightOptions,
    transformCSS,
    sortUtilities,
    safelist,
  }
}

export type ResolvedOptions = ReturnType<typeof resolveOptions>
