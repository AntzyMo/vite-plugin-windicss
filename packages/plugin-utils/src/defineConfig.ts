import { WindiCssOptions } from './options'

export function defineConfig(config: WindiCssOptions) {
  return config
}
