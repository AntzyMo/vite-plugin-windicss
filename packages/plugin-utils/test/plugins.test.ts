import { createBox } from '../src'

describe('plugins', () => {
  it('aspect-ratio', async() => {
    const box = createBox({
      config: {
        plugins: [
          require('windicss/plugin/aspect-ratio'),
        ],
      },
      scan: false,
      preflight: false,
    })
    box.init()
    box.extractFile('"aspect-none aspect-w-16 aspect-h-9 aspect-16/9"')

    const css = await box.generateCSS()
    expect(box.classesGenerated).toMatchSnapshot('classes')
    expect(css).toMatchSnapshot('generated-css')
  })

  it('forms', async() => {
    const box = createBox({
      config: {
        plugins: [
          require('windicss/plugin/forms'),
        ],
      },
      scan: false,
      preflight: {
        enableAll: true,
      },
    })
    box.init()

    const css = await box.generateCSS()
    expect(box.classesGenerated).toMatchSnapshot('classes')
    expect(css).toMatchSnapshot('generated-css')
  })
})
