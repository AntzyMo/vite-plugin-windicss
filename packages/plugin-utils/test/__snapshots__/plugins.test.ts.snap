// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins aspect-ratio: classes 1`] = `
Set {
  "aspect-none",
  "aspect-w-16",
  "aspect-h-9",
  "aspect-16/9",
}
`;

exports[`plugins aspect-ratio: generated-css 1`] = `
".aspect-none {
  position: static;
  padding-bottom: 0;
}
.aspect-none > * {
  position: static;
  height: auto;
  width: auto;
  top: auto;
  right: auto;
  bottom: auto;
  left: auto;
}
.aspect-w-16 {
  --tw-aspect-w: 16;
  position: relative;
  padding-bottom: calc(var(--tw-aspect-h) / var(--tw-aspect-w) * 100%);
}
.aspect-w-16 > * {
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}
.aspect-h-9 {
  --tw-aspect-h: 9;
}
.aspect-16\\\\/9 {
  position: relative;
  padding-bottom: 177.777778%;
}
.aspect-16\\\\/9 > * {
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}"
`;

exports[`plugins forms: classes 1`] = `Set {}`;

exports[`plugins forms: generated-css 1`] = `
"
[type='text'],
[multiple],
select
 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: #fff;
  border-color: #6b7280;
  border-width: 1px;
  border-radius: 0px;
  padding-top: 0.5rem;
  padding-right: 0.75rem;
  padding-bottom: 0.5rem;
  padding-left: 0.75rem;
  font-size: 1rem;
  line-height: 1.5rem;
}
[type='text']:focus, [multiple]:focus, select:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
  --tw-ring-inset: var(--tw-empty,/*!*/ /*!*/);
  --tw-ring-offset-width: 0px;
  --tw-ring-offset-color: #fff;
  --tw-ring-color: #2563eb;
  --tw-ring-offset-shadow: var(--tw-ring-inset) 0 0 0 var(--tw-ring-offset-width) var(--tw-ring-offset-color);
  --tw-ring-shadow: var(--tw-ring-inset) 0 0 0 calc(1px + var(--tw-ring-offset-width)) var(--tw-ring-color);
  box-shadow: var(--tw-ring-offset-shadow), var(--tw-ring-shadow), var(--tw-shadow, 0 0 #0000);
  border-color: #2563eb;
}
*, ::before, ::after {
  -webkit-box-sizing: border-box;
  box-sizing: border-box;
  border-width: 0;
  border-style: solid;
  border-color: #e5e7eb;
}
* {
  --tw-ring-inset: var(--tw-empty,/*!*/ /*!*/);
  --tw-ring-offset-width: 0px;
  --tw-ring-offset-color: #fff;
  --tw-ring-color: rgba(59, 130, 246, 0.5);
  --tw-ring-offset-shadow: 0 0 #0000;
  --tw-ring-shadow: 0 0 #0000;
  --tw-shadow: 0 0 #0000;
}
:root {
  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;
}
:-moz-focusring {
  outline: 1px dotted ButtonText;
}
:-moz-ui-invalid {
  box-shadow: none;
}
::moz-focus-inner {
  border-style: none;
  padding: 0;
}
::-webkit-inner-spin-button, ::-webkit-outer-spin-button {
  height: auto;
}
::-webkit-search-decoration {
  -webkit-appearance: none;
}
::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}
::-webkit-datetime-edit-fields-wrapper {
  padding: 0;
}
::-webkit-date-and-time-value {
  min-height: 1.5em;
}
[type='search'] {
  -webkit-appearance: textfield;
  outline-offset: -2px;
}
[multiple] {
  background-image: initial;
  background-position: initial;
  background-repeat: unset;
  background-size: initial;
  padding-right: 0.75rem;
  -webkit-print-color-adjust: unset;
  color-adjust: unset;
}
abbr[title] {
  -webkit-text-decoration: underline dotted;
  text-decoration: underline dotted;
}
body {
  margin: 0;
  font-family: inherit;
  line-height: inherit;
}
html {
  -webkit-text-size-adjust: 100%;
  font-family: ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,\\"Segoe UI\\",Roboto,\\"Helvetica Neue\\",Arial,\\"Noto Sans\\",sans-serif,\\"Apple Color Emoji\\",\\"Segoe UI Emoji\\",\\"Segoe UI Symbol\\",\\"Noto Color Emoji\\";
  line-height: 1.5;
}
input, select {
  font-family: inherit;
  font-size: 100%;
  line-height: 1.15;
  margin: 0;
  padding: 0;
  line-height: inherit;
  color: inherit;
}
input::placeholder {
  opacity: 1;
  color: #9ca3af;
  color: #6b7280;
}
input::webkit-input-placeholder {
  opacity: 1;
  color: #9ca3af;
}
input::-moz-placeholder {
  opacity: 1;
  color: #9ca3af;
  color: #6b7280;
}
input:-ms-input-placeholder {
  opacity: 1;
  color: #9ca3af;
}
input::-ms-input-placeholder {
  opacity: 1;
  color: #9ca3af;
  color: #6b7280;
}
select {
  text-transform: none;
  background-image: url(\\"data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' fill='none' viewBox='0 0 20 20'%3e%3cpath stroke='%236b7280' stroke-linecap='round' stroke-linejoin='round' stroke-width='1.5' d='M6 8l4 4 4-4'/%3e%3c/svg%3e\\");
  background-position: right 0.5rem center;
  background-repeat: no-repeat;
  background-size: 1.5em 1.5em;
  padding-right: 2.5rem;
  -webkit-print-color-adjust: exact;
  color-adjust: exact;
}"
`;
